src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py

```diff
@@ -0,0 +1,451 @@
+        "  enddo\n"
+        "  result = res_dot_product\n\n")
+    check_trans(code, expected, fortran_reader, fortran_writer, tmpdir)
+
+
+def test_apply_unknown_dims(tmpdir, fortran_reader, fortran_writer):
+    '''Test that the DotProduct2CodeTrans apply method produces the
+    expected PSyIR when neither of the vectors have a known size.
+
+    '''
+    code = (
+        "subroutine dot_product_test(v1,v2)\n"
+        "real,intent(in) :: v1(:), v2(:)\n"
+        "real :: result\n"
+        "result = dot_product(v1,v2)\n"
+        "end subroutine\n")
+    expected = (
+        "  integer :: i\n"
+        "  real :: res_dot_product\n\n"
+        "  res_dot_product = 0.0\n"
+        "  do i = LBOUND(v1, 1), UBOUND(v1, 1), 1\n"
+        "    res_dot_product = res_dot_product + v1(i) * v2(i)\n"
+        "  enddo\n"
+        "  result = res_dot_product\n\n")
+    check_trans(code, expected, fortran_reader, fortran_writer, tmpdir)
+
+
+def test_apply_multi_rhs(tmpdir, fortran_reader, fortran_writer):
+    '''Test that the DotProduct2CodeTrans apply method produces the
+    expected PSyIR when the expression on the rhs contains more than
+    just the DOT_PRODUCT.
+
+    '''
+    code = (
+        "subroutine dot_product_test(v1,v2)\n"
+        "real,intent(in) :: v1(10), v2(:)\n"
+        "real :: a, b, c, result\n"
+        "result = a + b*dot_product(v1,v2) + c\n"
+        "end subroutine\n")
+    expected = (
+        "  integer :: i\n"
+        "  real :: res_dot_product\n\n"
+        "  res_dot_product = 0.0\n"
+        "  do i = 1, 10, 1\n"
+        "    res_dot_product = res_dot_product + v1(i) * v2(i)\n"
+        "  enddo\n"
+        "  result = a + b * res_dot_product + c\n\n")
+    check_trans(code, expected, fortran_reader, fortran_writer, tmpdir,
+                index=3)
+
+
+@pytest.mark.parametrize("arg1,arg2", [("", "(:)"), ("(:)", ""),
+                                       ("(:)", "(:)")])
+def test_apply_array_notation(
+        tmpdir, fortran_reader, fortran_writer, arg1, arg2):
+    '''Test that the DotProduct2CodeTrans apply method produces the
+    expected PSyIR when array notation is used on the rhs by one or
+    both of the dot_product arguments.
+
+    '''
+    code = (
+        f"subroutine dot_product_test(v1,v2)\n"
+        f"real,intent(in) :: v1(:), v2(:)\n"
+        f"real :: result\n"
+        f"result = dot_product(v1{arg1},v2{arg2})\n"
+        f"end subroutine\n")
+    expected = (
+        "  integer :: i\n"
+        "  real :: res_dot_product\n\n"
+        "  res_dot_product = 0.0\n"
+        "  do i = LBOUND(v1, 1), UBOUND(v1, 1), 1\n"
+        "    res_dot_product = res_dot_product + v1(i) * v2(i)\n"
+        "  enddo\n"
+        "  result = res_dot_product\n\n")
+    check_trans(code, expected, fortran_reader, fortran_writer, tmpdir)
+
+
+@pytest.mark.parametrize("arg1,arg2,res1,res2",
+                         [("(:,n)", "(:,1)", "(i,n)", "(i,1)")])
```

Comment:
As commented earlier, please could you extend this test to cases where the two arrays have different numbers of dimensions.

Reply:
I've created an additional test `test_apply_extra_dims_sizes` that does this.